torch/distributed/_functional_collectives.py

```diff
@@ -375,6 +408,41 @@ def reduce_scatter_tensor(
     res = _maybe_wrap_tensor(tensor)
     return res
 
+
+def all_reduce_coalesced(self: List[torch.Tensor], reduceOp: str, group: RANK_TYPES, tag: str = "") -> List[torch.Tensor]:
+    """
+    Reduces a list of tensors across all machines in such a way that all get
+    the final result.
+
+    The all tensors in the input list are left unmodified.
+
+    Group can be one of:
+        List[int]: ranks participating in the collective.
+        List[List[int]]: 2D mesh of ranks taking part of this collective in MPMD.
+        ProcessGroup: Will perform a collective using the ranks and tag of the PG.
+        DeviceMesh: Do a SPMD collective over all ranks of the mesh
+        (DeviceMesh, int): Do a MPMD collective over one dimension of the DeviceMesh
+
+    :: N.B. If you pass a PG or a 1D list to perform a MPMD collective, the compiler won't be able to recover
+    that information and perform collective algebraic optimization. Use other forms of input for that.
+    """
+    tag, rankset, group_size = _expand_group(group, tag)
+    tensor_list = torch.ops.c10d_functional.all_reduce_coalesced(self, reduceOp, tag, rankset, group_size)  # type: ignore[attr-defined]
+
+    if _are_we_tracing():
+        return list(map(wait_tensor, tensor_list))
+
+    res = []
+    lookup_tensor = tensor_list[0]
+    wh = WaitHolder(lookup_tensor)
+    for tensor in tensor_list:
+        act = AsyncCollectiveTensor(tensor, wh)
+        res.append(cast(torch.Tensor, act))
+
+    # FIXME we should register all tensors and ref count when to emit the wait
+    _register_wrapper_tensor(res[0], lookup_tensor)
```

Comment:
shall we fix this inside this PR? i.e. maybe we should make the `AsyncCollectiveTensor` or `aten.wait` works on either a `tensor` or `list_of_tensors`, so that we only need to emit one `wait` call and that `wait` call should wait all tensors inside the list